File: reference-implementation/typescript/src/lib/tick-engine.ts
import { nowIso } from "./helpers.js";

export interface CandleState {
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  openTime: string;
  complete: boolean;
}

export interface TickEngineCallbacks {
  onQuoteUpdate(mid: number, bid: number, ask: number): void;
  onCandleUpdate(timeframe: string): void;
  onCandleClose(timeframe: string, candle: CandleState): void;
  onFeatureUpdate(): void;
}

const TIMEFRAMES = ["M1", "M5", "H1"] as const;
type Timeframe = (typeof TIMEFRAMES)[number];

const HALF_SPREAD = 0.0001;
const MAX_PIP_STEP = 0.0002;
const HISTORY_LIMIT = 300;
const TICK_INTERVAL_MS = 2_000;

function roundTo5(value: number): number {
  return Math.round(value * 100_000) / 100_000;
}

export class TickEngine {
  private mid = 1.0875;
  private priceHistory: number[] = [];
  private candles = new Map<Timeframe, CandleState>();
  private intervalId: ReturnType<typeof setInterval> | null = null;
  private callbacks: TickEngineCallbacks;

  constructor(callbacks: TickEngineCallbacks) {
    this.callbacks = callbacks;
    for (const tf of TIMEFRAMES) {
      this.candles.set(tf, this.freshCandle(this.mid));
    }
  }

  start(): void {
    if (this.intervalId !== null) return;
    this.intervalId = setInterval(() => this.tick(), TICK_INTERVAL_MS);
  }

  stop(): void {
    if (this.intervalId !== null) {
      clearInterval(this.intervalId);
      this.intervalId = null;
    }
  }

  forceCandleClose(timeframe: string): void {
    const tf = timeframe as Timeframe;
    const candle = this.candles.get(tf);
    if (!candle) return;
    candle.complete = true;
    this.callbacks.onCandleClose(tf, { ...candle });
    this.candles.set(tf, this.freshCandle(this.mid));
  }

  getCandle(timeframe: string): CandleState | undefined {
    return this.candles.get(timeframe as Timeframe);
  }

  getMid(): number {
    return this.mid;
  }

  getReturns(): { r_1s: number; r_5s: number; r_1m: number } {
    const len = this.priceHistory.length;
    if (len < 2) {
      return { r_1s: 0, r_5s: 0, r_1m: 0 };
    }

    const current = this.priceHistory[len - 1];
    // Each tick is 2 seconds apart.
    // r_1s: 1 tick back (closest we have to 1s at 2s interval)
    const r_1s = len >= 2
      ? (current - this.priceHistory[len - 2]) / this.priceHistory[len - 2]
      : 0;
    // r_5s: ~3 ticks back (6s worth, closest to 5s)
    const idx5s = Math.max(0, len - 3);
    const r_5s = (current - this.priceHistory[idx5s]) / this.priceHistory[idx5s];
    // r_1m: ~30 ticks back (60s)
    const idx1m = Math.max(0, len - 30);
    const r_1m = (current - this.priceHistory[idx1m]) / this.priceHistory[idx1m];

    return {
      r_1s: roundTo5(r_1s),
      r_5s: roundTo5(r_5s),
      r_1m: roundTo5(r_1m),
    };
  }

  getVolatility(): { rv_1m: number; rv_5m: number } {
    const len = this.priceHistory.length;
    if (len < 3) {
      return { rv_1m: 0, rv_5m: 0 };
    }

    // Compute log returns over the available history
    const logReturns: number[] = [];
    for (let i = 1; i < len; i++) {
      logReturns.push(Math.log(this.priceHistory[i] / this.priceHistory[i - 1]));
    }

    // rv_1m: variance of last ~30 ticks (60s) annualized
    // rv_5m: variance of last ~150 ticks (300s) annualized
    const rv_1m = this.annualizedVol(logReturns.slice(-30));
    const rv_5m = this.annualizedVol(logReturns.slice(-150));

    return {
      rv_1m: roundTo5(rv_1m),
      rv_5m: roundTo5(rv_5m),
    };
  }

  private annualizedVol(returns: number[]): number {
    if (returns.length < 2) return 0;
    const mean = returns.reduce((s, v) => s + v, 0) / returns.length;
    const variance = returns.reduce((s, v) => s + (v - mean) ** 2, 0) / (returns.length - 1);
    // Each observation is 2 seconds apart.
    // Observations per year = 252 days * 24h * 3600s / 2s = 10,886,400
    const obsPerYear = 252 * 24 * 3600 / 2;
    return Math.sqrt(variance * obsPerYear);
  }

  private tick(): void {
    // Random walk
    const delta = (Math.random() - 0.5) * 2 * MAX_PIP_STEP;
    this.mid = roundTo5(this.mid + delta);

    const bid = roundTo5(this.mid - HALF_SPREAD);
    const ask = roundTo5(this.mid + HALF_SPREAD);

    // Push to history, cap at HISTORY_LIMIT
    this.priceHistory.push(this.mid);
    if (this.priceHistory.length > HISTORY_LIMIT) {
      this.priceHistory.shift();
    }

    // Update all candle states
    for (const tf of TIMEFRAMES) {
      const candle = this.candles.get(tf)!;
      candle.high = Math.max(candle.high, this.mid);
      candle.low = Math.min(candle.low, this.mid);
      candle.close = this.mid;
      candle.volume += 1;
    }

    // Check wall-clock boundaries for candle closes
    const now = new Date();
    const seconds = now.getSeconds();
    const minutes = now.getMinutes();

    if (seconds < 2) {
      const m1Candle = this.candles.get("M1")!;
      if (m1Candle.volume > 0) {
        this.forceCandleClose("M1");
      }

      if (minutes % 5 === 0) {
        const m5Candle = this.candles.get("M5")!;
        if (m5Candle.volume > 0) {
          this.forceCandleClose("M5");
        }
      }

      if (minutes === 0) {
        const h1Candle = this.candles.get("H1")!;
        if (h1Candle.volume > 0) {
          this.forceCandleClose("H1");
        }
      }
    }

    // Fire callbacks
    this.callbacks.onQuoteUpdate(this.mid, bid, ask);
    this.callbacks.onCandleUpdate("M1");
    this.callbacks.onFeatureUpdate();
  }

  private freshCandle(price: number): CandleState {
    return {
      open: price,
      high: price,
      low: price,
      close: price,
      volume: 0,
      openTime: nowIso(),
      complete: false,
    };
  }
}

File: reference-implementation/go/tick_engine.go
package main

import (
	"math"
	"math/rand"
	"sync"
	"time"
)

const (
	halfSpread   = 0.0001
	maxPipStep   = 0.0002
	historyLimit = 300
	tickInterval = 2 * time.Second
)

// CandleState holds the OHLCV data for a partial or completed candle.
type CandleState struct {
	Open     float64
	High     float64
	Low      float64
	Close    float64
	Volume   int
	OpenTime string
	Complete bool
}

// TickEngineCallbacks defines the callbacks fired by the tick engine.
type TickEngineCallbacks struct {
	OnQuoteUpdate   func(mid, bid, ask float64)
	OnCandleUpdate  func(timeframe string)
	OnCandleClose   func(timeframe string, candle CandleState)
	OnFeatureUpdate func()
}

// TickEngine simulates market data by random-walking a mid price.
type TickEngine struct {
	mu           sync.Mutex
	mid          float64
	priceHistory []float64
	candles      map[string]*CandleState
	callbacks    TickEngineCallbacks
	ticker       *time.Ticker
	stopCh       chan struct{}
	running      bool
}

// NewTickEngine creates a new tick engine with the given callbacks.
func NewTickEngine(callbacks TickEngineCallbacks) *TickEngine {
	te := &TickEngine{
		mid:          1.0875,
		priceHistory: []float64{},
		candles:      map[string]*CandleState{},
		callbacks:    callbacks,
		stopCh:       make(chan struct{}),
	}
	for _, tf := range []string{"M1", "M5", "H1"} {
		te.candles[tf] = te.freshCandle(te.mid)
	}
	return te
}

// Start begins the tick engine's periodic price updates.
func (te *TickEngine) Start() {
	te.mu.Lock()
	if te.running {
		te.mu.Unlock()
		return
	}
	te.running = true
	te.ticker = time.NewTicker(tickInterval)
	te.mu.Unlock()

	go func() {
		for {
			select {
			case <-te.ticker.C:
				te.tick()
			case <-te.stopCh:
				return
			}
		}
	}()
}

// Stop halts the tick engine.
func (te *TickEngine) Stop() {
	te.mu.Lock()
	defer te.mu.Unlock()
	if !te.running {
		return
	}
	te.running = false
	te.ticker.Stop()
	close(te.stopCh)
}

// ForceCandleClose closes the current partial candle for the given timeframe.
func (te *TickEngine) ForceCandleClose(timeframe string) {
	te.mu.Lock()
	candle, ok := te.candles[timeframe]
	if !ok {
		te.mu.Unlock()
		return
	}
	candle.Complete = true
	closed := *candle
	te.candles[timeframe] = te.freshCandle(te.mid)
	te.mu.Unlock()

	if te.callbacks.OnCandleClose != nil {
		te.callbacks.OnCandleClose(timeframe, closed)
	}
}

// GetMid returns the current mid price.
func (te *TickEngine) GetMid() float64 {
	te.mu.Lock()
	defer te.mu.Unlock()
	return te.mid
}

// GetReturns computes simple returns over recent price history.
func (te *TickEngine) GetReturns() map[string]float64 {
	te.mu.Lock()
	defer te.mu.Unlock()

	l := len(te.priceHistory)
	if l < 2 {
		return map[string]float64{"r_1s": 0, "r_5s": 0, "r_1m": 0}
	}

	current := te.priceHistory[l-1]

	r1s := 0.0
	if l >= 2 {
		r1s = (current - te.priceHistory[l-2]) / te.priceHistory[l-2]
	}

	idx5s := l - 3
	if idx5s < 0 {
		idx5s = 0
	}
	r5s := (current - te.priceHistory[idx5s]) / te.priceHistory[idx5s]

	idx1m := l - 30
	if idx1m < 0 {
		idx1m = 0
	}
	r1m := (current - te.priceHistory[idx1m]) / te.priceHistory[idx1m]

	return map[string]float64{
		"r_1s": roundTo5(r1s),
		"r_5s": roundTo5(r5s),
		"r_1m": roundTo5(r1m),
	}
}

// GetVolatility computes annualized realized volatility.
func (te *TickEngine) GetVolatility() map[string]float64 {
	te.mu.Lock()
	defer te.mu.Unlock()

	l := len(te.priceHistory)
	if l < 3 {
		return map[string]float64{"rv_1m": 0, "rv_5m": 0}
	}

	logReturns := make([]float64, 0, l-1)
	for i := 1; i < l; i++ {
		logReturns = append(logReturns, math.Log(te.priceHistory[i]/te.priceHistory[i-1]))
	}

	rv1m := annualizedVol(lastN(logReturns, 30))
	rv5m := annualizedVol(lastN(logReturns, 150))

	return map[string]float64{
		"rv_1m": roundTo5(rv1m),
		"rv_5m": roundTo5(rv5m),
	}
}

// GetCandle returns the current candle state for a timeframe.
func (te *TickEngine) GetCandle(timeframe string) *CandleState {
	te.mu.Lock()
	defer te.mu.Unlock()
	c, ok := te.candles[timeframe]
	if !ok {
		return nil
	}
	copy := *c
	return &copy
}

func (te *TickEngine) tick() {
	te.mu.Lock()

	// Random walk
	delta := (rand.Float64() - 0.5) * 2 * maxPipStep
	te.mid = roundTo5(te.mid + delta)

	bid := roundTo5(te.mid - halfSpread)
	ask := roundTo5(te.mid + halfSpread)

	te.priceHistory = append(te.priceHistory, te.mid)
	if len(te.priceHistory) > historyLimit {
		te.priceHistory = te.priceHistory[1:]
	}

	// Update all candle states
	for _, tf := range []string{"M1", "M5", "H1"} {
		candle := te.candles[tf]
		if te.mid > candle.High {
			candle.High = te.mid
		}
		if te.mid < candle.Low {
			candle.Low = te.mid
		}
		candle.Close = te.mid
		candle.Volume++
	}

	// Check wall-clock boundaries for candle closes
	now := time.Now()
	seconds := now.Second()
	minutes := now.Minute()

	var candlesToClose []string

	if seconds < 2 {
		m1Candle := te.candles["M1"]
		if m1Candle.Volume > 0 {
			candlesToClose = append(candlesToClose, "M1")
		}

		if minutes%5 == 0 {
			m5Candle := te.candles["M5"]
			if m5Candle.Volume > 0 {
				candlesToClose = append(candlesToClose, "M5")
			}
		}

		if minutes == 0 {
			h1Candle := te.candles["H1"]
			if h1Candle.Volume > 0 {
				candlesToClose = append(candlesToClose, "H1")
			}
		}
	}

	mid := te.mid
	te.mu.Unlock()

	// Fire callbacks outside the lock
	for _, tf := range candlesToClose {
		te.ForceCandleClose(tf)
	}

	if te.callbacks.OnQuoteUpdate != nil {
		te.callbacks.OnQuoteUpdate(mid, bid, ask)
	}
	if te.callbacks.OnCandleUpdate != nil {
		te.callbacks.OnCandleUpdate("M1")
	}
	if te.callbacks.OnFeatureUpdate != nil {
		te.callbacks.OnFeatureUpdate()
	}
}

func (te *TickEngine) freshCandle(price float64) *CandleState {
	return &CandleState{
		Open:     price,
		High:     price,
		Low:      price,
		Close:    price,
		Volume:   0,
		OpenTime: nowISO(),
		Complete: false,
	}
}

func roundTo5(v float64) float64 {
	return math.Round(v*100000) / 100000
}

func annualizedVol(returns []float64) float64 {
	if len(returns) < 2 {
		return 0
	}
	sum := 0.0
	for _, r := range returns {
		sum += r
	}
	mean := sum / float64(len(returns))
	variance := 0.0
	for _, r := range returns {
		variance += (r - mean) * (r - mean)
	}
	variance /= float64(len(returns) - 1)
	// Each observation is 2 seconds apart.
	// Observations per year = 252 days * 24h * 3600s / 2s = 10,886,400
	obsPerYear := float64(252 * 24 * 3600 / 2)
	return math.Sqrt(variance * obsPerYear)
}

func lastN(slice []float64, n int) []float64 {
	if len(slice) <= n {
		return slice
	}
	return slice[len(slice)-n:]
}
